src/common/http/client.go

```diff
@@ -28,41 +28,80 @@ import (
 	"reflect"
 )
 
-// Client is a util for common HTTP operations, such Get, Head, Post, Put and Delete.
-// Use Do instead if  those methods can not meet your requirement
-type Client struct {
-	modifiers []modifier.Modifier
-	client    *http.Client
-}
+const (
+	// DefaultTransport used to get the default http Transport
+	DefaultTransport = iota
+	// InsecureTransport used to get the insecure http Transport
+	InsecureTransport
+	// InternalTransport used to get the internal secure http Transport
+	InternalTransport
+	// SecureTransport used to get the external secure http Transport
+	SecureTransport
+)
 
-var defaultHTTPTransport, secureHTTPTransport, insecureHTTPTransport *http.Transport
+var (
+	secureHTTPTransport   *http.Transport
+	insecureHTTPTransport *http.Transport
+	internalTransport     *http.Transport
+)
 
 func init() {
-	defaultHTTPTransport = &http.Transport{}
 
 	secureHTTPTransport = &http.Transport{
 		Proxy: http.ProxyFromEnvironment,
 		TLSClientConfig: &tls.Config{
 			InsecureSkipVerify: false,
 		},
 	}
+
 	insecureHTTPTransport = &http.Transport{
 		Proxy: http.ProxyFromEnvironment,
 		TLSClientConfig: &tls.Config{
 			InsecureSkipVerify: true,
 		},
 	}
+
+	initInternalTransport()
 }
 
-// GetHTTPTransport returns HttpTransport based on insecure configuration
-func GetHTTPTransport(insecure ...bool) *http.Transport {
-	if len(insecure) == 0 {
-		return defaultHTTPTransport
+// Client is a util for common HTTP operations, such Get, Head, Post, Put and Delete.
+// Use Do instead if  those methods can not meet your requirement
+type Client struct {
+	modifiers []modifier.Modifier
+	client    *http.Client
+}
+
+func initInternalTransport() {
+	if InternalTLSEnabled() {
+		tlsConfig, err := GetInternalTLSConfig()
+		if err != nil {
+			panic(err)
+		}
+		internalTransport = &http.Transport{
+			TLSClientConfig: tlsConfig,
+		}
+	} else {
+		internalTransport = &http.Transport{
+			TLSClientConfig: &tls.Config{
+				InsecureSkipVerify: true,
+			},
+		}
 	}
-	if insecure[0] {
-		return insecureHTTPTransport
+}
+
+// GetHTTPTransport returns HttpTransport based on insecure configuration
+func GetHTTPTransport(clientType uint) *http.Transport {
```

Comment:
Please test in mass replication scenario, I remember  each time you create a new transport there will be a new connection ,and it will be pending at "TIME_WAIT" until you set some timeout values.

Reply:
There is no new connection when you create a new Transport. The new connection only starts when you start a request with new Transport. In this situation. These changes will improve our concurrency in some bit. If it caused too many time_wait we can use a connection pool or config the kernel parameters. I think in many companies ops guy will config the kernel configuration to cut down the `TIME_WAIT` time.